TestVectors/Ed25519Signature2020/proofCanonEdSig.txt

```diff
@@ -0,0 +1,4 @@
+_:c14n0 <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://www.w3.org/ns/credentials/issuer-dependent#Ed25519Signature2020> .
```

Comment:
This is incorrect due to the missing `https://w3id.org/security/suites/ed25519-2020/v1` context.

Reply:
I'm not quite sure about all the @context fields currently in the draft. When I look at Examples 5 and 6 they just use "https://w3id.org/security/data-integrity/v1" and not "https://w3id.org/security/suites/ed25519-2020/v1". "https://w3id.org/security/data-integrity/v1" was subsumed into "https://www.w3.org/ns/credentials/v2".

The ed25519-2020 context contains some of the same fields. Shouldn't the draft tell us which contexts to use in each case?

Reply:
For the JCS version, there wasn't any text to go by. Since it is an extra step to put the `@context` back into the proof options and the `@context` isn't needed for JCS (JSON canonicalization scheme). I left out the step.

Reply:
Concerning: "There was one other difference that we will want to sort out. My understanding is that we wanted to use secretKeyMultibase over privateKeyMultibase (the spec says the latter today) -- and that we should probably use the 64 byte key representation instead of the 32 byte key representation. The 64 byte key representation just concatenates the public key value after the private key value, which is what most implementations, I believe, use."

I think @msporny raised a point like this before. I'm not sure that there will be a standard for an unwrapped private/secret key though [multicodec](https://ipfs.io/ipfs/QmXec1jjwzxWJoNbxQF5KffL8q6hFXm9QwUGaa3wKGk6dT/#title=Multicodecs&src=https://raw.githubusercontent.com/multiformats/multicodec/master/table.csv) lists some. Most documents with test vectors that include secret key just give the HEX. Maybe we should just do that.

I would find it really confusing to call something a private/secret key and then find its the wrong length because it has a public key appended.

Reply:
@Wind4Greg,

> I would find it really confusing to call something a private/secret key and then find its the wrong length because it has a public key appended.

Yes, that may be true, but I think that's just a consequence of a bigger problem: having key expressions that fail to encapsulate crypto primitive internal components and failing to have application layer code treat crypto primitive expressions as opaque blobs.

Some key expressions, such for some key types expressed as JWKs, suffer from this -- whereby individual components of cryptographic keys are delimited and exposed in the JSON expression of the key, rather than encapsulating everything in a blob that is handed to the crypto-layer implementation. I think this is a layering violation and a mistake that can result in a need for more validation code and other bugs / confusion / attack surface.

Ultimately, at this layer, we shouldn't have to care what's in the field -- just what the field means ("it's a 'secret key of type X'") and how the value is base-encoded. That's all we should need to decode the thing and pass it to the right crypto-layer implementation. Whether crypto implementations standardize on private key expressions that include the public key concatenated onto the end or not should be totally irrelevant to us at this layer -- we should treat these things as opaque and just pass them through.

Once we find ourselves validating whether some point is compressed to just the `x` coordinate and it can expand out the `y` coordinate or if the Chinese Remainder theorem components are present and work properly with one another -- we've failed at this layer, IMO.

So, at this layer, I think we should not care what the blob of bytes is for the private key -- we just know that's what it is. That blob should conform to what most implementations accept. Now, for whatever reason, that happens to be a 64 byte array with the private key seed + associated public key concatenated onto it. That this was the choice that most crypto-layer implementations made shouldn't matter to us, we should just say that's how it is and run with it, IMO.

Reply:
I think we are in agreement! I'm still not sure about what should be in the `@context` fields of examples (5 and 6) in the draft and the test vector unsigned document. Will we really need two different test vector unsigned base documents just differing in their `@context` fields? Maybe since one maybe for legacy. I'd like @msporny to chime in since he suggested the base test vector document. I can rerun test vectors relatively easily.

Reply:
Additional good rationale for the secret key to be the 64-byte concatenation of the secret seed and public key point: https://crypto.stackexchange.com/questions/54353/why-are-nacl-secret-keys-64-bytes-for-signing-but-32-bytes-for-box

So I think we should make clear that that is what it is in our specs.

Reply:
Sounds good.